### PR TITLE
Attempt to optimize dfa

### LINKED ISSUE #50: Change “if” to “if state”
"If" is too short, so sometimes gets recognised when it shouldn't

### PR DESCRIPTION
**Pull request by pokey**
I kept almost all functionality, but dropped a couple things I never use, and the grammar is maybe a bit more awkward

The biggest grammar change here is the unification of all the multi-target commands.  It's now "bring T onto T", "swap T onto T", "call T onto T", and "move T onto T", rather than each having a different preposition.  In each case now both "T"s are optional, so you can eg "bring onto T" etc

Curious what people think

One of my takeaways here is that DFA compilation seems to be quicker if big captures appear towards the end of a command rather than the beginning.  @lunixbochs does that sound right to you?

- [x] "replace T with X" should change to "change T word X".  We no longer support arbitrary format strings here.  We can then add "add" with same syntax "add T word X", but by default will modify T to have an "after".  In the future, it will be its own action (see https://github.com/pokey/cursorless-vscode/issues/210)
- [x] change the recursive descent code for “after”
- [x]  it’s "[inner | outer | pair] {enclosingScopeType} | (inner | outer | pair)"
- [x] And remove insideOutsideType
- [x] And switch “tween” to “between”
- [x] And switch bringSwapMove grammar back

fixes #50

### CODE REVIEW

File: src/actions/multiple_target_action.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+mod = Module()
+
+ctx = Context()
+ctx.matches = r"""
+tag: user.cursorless
+"""
+
+
+mod.list("cursorless_multiple_target_action", desc="Cursorless move or bring actions")
+ctx.lists["self.cursorless_multiple_target_action"] = {"bring", "move", "swap", "call"}
```
Comment (pokey) [inline src/actions/multiple_target_action.py:13]: Unified the syntax for all these multi-target commands. Slightly awkward for some of them but not too bad so far; will have to see how well it works for a couple days

File: src/compound_targets.py [hunk 2]
```diff
@@ -1,14 +1,27 @@
 from .primitive_target import BASE_TARGET
-from talon import Module
+from talon import Module, Context
```
Comment (pokey) [inline src/compound_targets.py:2]: no actual syntax changes here

File: src/cursorless.talon [hunk 3]
```diff
@@ -5,32 +5,26 @@ tag(): user.cursorless
...
 
+{user.cursorless_multiple_target_action} <user.cursorless_multiple_targets>:
+    user.cursorless_multiple_target_command(cursorless_multiple_target_action, cursorless_multiple_targets)
+
 <user.cursorless_wrapper> wrap <user.cursorless_target>:
     user.cursorless_single_target_command_with_arg_list("wrap", cursorless_target, cursorless_wrapper)
 
-wrap <user.cursorless_target> with funk <user.code_functions>:
-    user.cursorless_single_target_command("wrap", cursorless_target, "{code_functions}(", ")")
+# wrap <user.cursorless_target> with funk <user.code_functions>:
```
Comment (pokey) [inline src/cursorless.talon:14]: I never use this

File: src/cursorless.talon [hunk 4]
```diff
@@ -5,32 +5,26 @@ tag(): user.cursorless
...
-
-<user.cursorless_swap>:
-    user.cursorless_multiple_target_command("swap", cursorless_swap)
-
-{user.cursorless_move_bring_action} <user.cursorless_move_bring_targets>:
-    user.cursorless_multiple_target_command(cursorless_move_bring_action, cursorless_move_bring_targets)
-
-<user.cursorless_call>:
-    user.cursorless_multiple_target_command("call", cursorless_call)
+# extract <user.cursorless_target> as <user.text>$:
```
Comment (pokey) [inline src/cursorless.talon:23]: or this

File: src/cursorless.talon [hunk 5]
```diff
@@ -5,32 +5,26 @@ tag(): user.cursorless
...
-wrap <user.cursorless_target> with funk <user.code_functions>:
-    user.cursorless_single_target_command("wrap", cursorless_target, "{code_functions}(", ")")
+# wrap <user.cursorless_target> with funk <user.code_functions>:
+#     user.cursorless_single_target_command("wrap", cursorless_target, "{code_functions}(", ")")
 
 replace <user.cursorless_target> with <user.cursorless_replace_value>$:
     user.cursorless_replace(cursorless_target, cursorless_replace_value)
 
-reformat <user.cursorless_target> as <user.formatters>:
+format <user.formatters> onto <user.cursorless_target>:
```
Comment (pokey) [inline src/cursorless.talon:20]: ok so this one is pretty awkward, but surprisingly swapping these two captures seems to improve DFA compilation.  We could prob come up with better terminology here

File: src/modifiers/surrounding_pair.py [hunk 6]
```diff
@@ -8,7 +8,6 @@
 """
 
 pair_symbols = {
-    "angle": "angleBrackets",
```
Comment (pokey) [inline src/modifiers/surrounding_pair.py:11]: this was actually just causing problems from previous merge; "angle" is also the name of the char in my repo

File: src/primitive_target.py [hunk 7]
```diff
@@ -30,15 +30,10 @@
...
-    "<user.cursorless_mark>"  # 1 mark
-)
-
-modifiers_only = f"({'|'.join(modifiers)})+"  # 1 or more parameters
 
-
-@mod.capture(rule=f"({modifiers_and_mark}) | ({modifiers_only})")
+@mod.capture(
+    rule=f"({'|'.join(modifiers)})+ [<user.cursorless_mark>] | <user.cursorless_mark>"
+)
```
Comment (pokey) [inline src/primitive_target.py:36]: No actual grammar change here, but did seem to improve times somewhat

**Comment by lunixbochs** [comment]
Talon can pre-optimize multiple commands that begin and end with the same capture in some cases. Maybe that's what you're seeing.

File: src/modifiers/surrounding_pair.py (hunk 6 above)
Comment (AndreasArvidsson) [reply src/modifiers/surrounding_pair.py:11]: Couldn't we just rename it to diamond? I think we probably should synchronize wrap and pair commands.

File: src/actions/multiple_target_action.py (hunk 1 above)
Comment (AndreasArvidsson) [reply src/actions/multiple_target_action.py:13]: I total understand why you did it but I'm not quite happy with it :s

File: src/modifiers/surrounding_pair.py (hunk 6 above)
Comment (pokey) [reply src/modifiers/surrounding_pair.py:11]: See the next line 😊

File: src/cursorless.talon (hunk 5 above)
Comment (AndreasArvidsson) [reply src/cursorless.talon:20]: Maybe we should just remove onto and make it `format camel blue air`

File: src/primitive_target.py [hunk 8]
```diff
@@ -19,31 +13,32 @@
...
-    "<user.cursorless_head_tail>",  # head, tail
-    "<user.cursorless_inside_outside>",  # inner, outer
-    "<user.cursorless_surrounding_pair>",  # curly, round
+    "<user.cursorless_surrounding_pair>",  # matching/pair [curly, round]
     # "<user.cursorless_matching_pair_symbol>",  # matching
 ]
 
 
+
+@mod.capture(rule=f"{'|'.join(modifiers)}")
```
Comment (pokey) [inline src/primitive_target.py:30]:
```suggestion
@mod.capture(rule='|'.join(modifiers))
```

File: src/modifiers/surrounding_pair.py [hunk 9]
```diff
@@ -7,31 +7,26 @@
...
-mod.list("cursorless_pair_symbol", desc="A symbol that comes in pairs, eg brackets")
-ctx.lists["self.cursorless_pair_symbol"] = pair_symbols
 
-
-@mod.capture(
-    rule=(
-        "matching | {user.cursorless_pair_symbol} | pair [{user.cursorless_pair_symbol}] "
-    )
-)
+@mod.capture(rule=("(matching | pair) [{user.cursorless_pair_symbol}]"))
```
Comment (pokey) [inline src/modifiers/surrounding_pair.py:23]: I'd be inclined to revert this one

File: src/modifiers/head_tail.py [hunk 10]
```diff
@@ -1,8 +1,16 @@
...
 
 mod = Module()
+ctx = Context()
 
+ctx.matches = r"""
+tag: user.cursorless
+"""
 
-@mod.capture(rule="head | tail")
+mod.list("cursorless_head_tail", desc="Types of head_tail")
```
Comment (pokey) [inline src/modifiers/head_tail.py:10]:
```suggestion
mod.list("cursorless_head_tail", desc="Cursorless modifier for head or tail of line")
```

File: src/modifiers/containing_scope.py [hunk 11]
```diff
@@ -8,7 +7,9 @@
 tag: user.cursorless
 """
 
-containing_scope_type_map = {
+
+mod.list("cursorless_containing_scope_type", desc="Supported containing scope types")
```
Comment (pokey) [inline src/modifiers/containing_scope.py:11]: I'll rename this one to
```suggestion
mod.list("cursorless_scope_type", desc="Supported scope types")
```

as we'll use these for other things later

File: src/actions/multiple_target_action.py [hunk 12]
```diff
@@ -35,6 +34,8 @@ def cursorless_multiple_targets(m) -> str:
 
     try:
         target_separater = m.cursorless_target_separator
+        if target_separater != "at":
+            target_list[-1]["position"] = target_separater
```
Comment (pokey) [inline src/actions/multiple_target_action.py:38]: We'll need to fix this one; won't work for list targets

Comment (AndreasArvidsson) [reply src/actions/multiple_target_action.py:38]: Ah damn I hate weakly typed languages. Probably should make a utility for updating single or list of targets.

Comment (AndreasArvidsson) [reply src/actions/multiple_target_action.py]: @pokey I have pushed a change which updates lists but I have no idea how I'm going to add positions to a range target. Whatever I try the result is incorrect.

Comment (pokey) [reply src/actions/multiple_target_action.py]: Don't worry bout range targets.  Positional modifiers don't really make sense for them so do whatever you want

Comment (AndreasArvidsson) [reply src/actions/multiple_target_action.py]: Ok then Im done with this change. Pushed and ready.

File: src/actions/multiple_target_action.py [hunk 13]
```diff
@@ -46,3 +39,15 @@ def cursorless_multiple_targets(m) -> str:
...
+
+
+def update_target(target: dict, value: dict):
+    type = target["type"]
+    if type == "list":
+        for target in target["elements"]:
+            update_target(target, value)
+    # elif type == "range":
+        # update_target(target["start"], value)
+        # update_target(target["end"], value)
```
Comment (pokey) [inline src/actions/multiple_target_action.py:51]: for the generic function, I think this is probably the right behaviour.  Shall we keep it?

Comment (pokey) [reply src/actions/multiple_target_action.py:51]: actually maybe we should just operate on first thing and then let inference do the rest of the work

Comment (pokey) [reply src/actions/multiple_target_action.py:51]: so if list, operate on first list element, if range operate on start

Comment (pokey) [reply src/actions/multiple_target_action.py:51]: and then i think extension-side inference will do its work from there

Comment (pokey) [reply src/actions/multiple_target_action.py:51]: k i'll take this one over

File: src/modifiers/surrounding_pair.py [hunk 14]
```diff
@@ -7,36 +7,43 @@
 tag: user.cursorless
 """
 
-pair_symbols = {
-    "angle": "angleBrackets",
-    "diamond": "angleBrackets",
+
+mod.list("cursorless_pair_symbol", desc="A symbol that comes in pairs, eg brackets")
+ctx.lists["self.cursorless_pair_symbol"] = {
+    "skis": "backtickQuotes",
```
Comment (AndreasArvidsson) [inline src/modifiers/surrounding_pair.py:13]: Can't we make do with just `brick wrap`?

**Comment by AndreasArvidsson** [comment]
Can we change `paste` to something else?

File: src/actions/actions.py [hunk 15]
```diff
@@ -20,8 +20,8 @@ class MakeshiftAction:
 
 
 makeshift_actions = [
-    MakeshiftAction("define", "editor.action.revealDefinition"),
     MakeshiftAction("drink cell", "jupyter.insertCellAbove"),
+    MakeshiftAction("follow", "editor.action.revealDefinition"),
```
Comment (AndreasArvidsson) [inline src/actions/actions.py:24]: Personally I use reveal definition much more than I use find. Maybe we should rename find instead?

**Review by AndreasArvidsson — APPROVED** [review]

**Pull request merged**